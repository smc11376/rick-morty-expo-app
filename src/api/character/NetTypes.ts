export type CharacterNetType = {
  id: number;
  name: string;
  status: string;
  species: string;
  type: string;
  gender: string;
  origin: OriginOrLocationNetType;
  location: OriginOrLocationNetType;
  image: string;
  episode?: string[] | null;
  url: string;
  created: string;
};

export type OriginOrLocationNetType = {
  name: string;
  url: string;
};

export type GetAllCharactersNetType = {
  info: {
    count: number;
    pages: number;
    next: string | null;
    prev: string | null;
  };
  results: CharacterNetType[];
};
