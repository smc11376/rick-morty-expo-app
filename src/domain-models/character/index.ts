export type ItemWithUrl = {
  name: string;
  url: string;
};

export type ShortCharacter = {
  id: number;
  name: string;
  status: string;
  image: string;
};

export type Character = ShortCharacter & {
  species: string;
  type: string;
  gender: string;
  origin: ItemWithUrl;
  location: ItemWithUrl;
  numberOfEpisodes: number;
};
