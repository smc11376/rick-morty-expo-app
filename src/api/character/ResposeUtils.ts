import { CharacterNetType, GetAllCharactersNetType } from "@/api/character/NetTypes";
import { Character, ShortCharacter } from "@/domain-models/character";

export function transformGetAllCharactersResponse(response?: GetAllCharactersNetType): ShortCharacter[] {
  if (!response) return [];

  return response.results.map(item => ({
    id: item.id,
    name: item.name,
    status: item.status,
    image: item.image,
  }));
}

export function transformGetCharacterByIdResponse(response?: CharacterNetType): Character | undefined {
  if (!response) return undefined;

  return {
    id: response.id,
    name: response.name,
    status: response.status,
    species: response.species,
    type: response.type,
    gender: response.gender,
    origin: response.origin,
    location: response.location,
    image: response.image,
    numberOfEpisodes: response.episode?.length || 0,
  };
}
