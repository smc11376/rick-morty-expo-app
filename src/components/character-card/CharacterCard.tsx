import { Image, StyleSheet, Text, TouchableOpacity, View } from "react-native";
import { ShortCharacter } from "@/domain-models/character";
import { getStatusColor } from "@/components/character-card/CharacterCard.Utils";

type CharacterCardProps = {
  character: ShortCharacter;
  onPressCharacter: () => void;
};

const CARD_HEIGHT = 80;
const CARD_WIDTH = 80;
const CARD_PADDING_VERTICAL = 8;
export const CARD_SIZE = CARD_HEIGHT + 2 * CARD_PADDING_VERTICAL;

export default function CharacterCard({ character, onPressCharacter }: CharacterCardProps) {
  const { name, image, status } = character;
  return (
    <TouchableOpacity testID="character-card-wrapper" style={styles.wrapper} onPress={onPressCharacter}>
      <View style={styles.container}>
        <Image style={styles.image} source={{ uri: image }} resizeMode="contain" />
        <View style={styles.infoContainer}>
          <Text style={styles.title}>{name}</Text>
          <Text style={styles.status}>
            {"Status: "}
            <Text style={getStatusColor(status)}>{status}</Text>
          </Text>
        </View>
      </View>
    </TouchableOpacity>
  );
}

const styles = StyleSheet.create({
  wrapper: {
    paddingHorizontal: 24,
    paddingVertical: CARD_PADDING_VERTICAL,
  },
  container: {
    flexDirection: "row",
    alignItems: "center",
    backgroundColor: "rgb(60, 62, 68)",
    minHeight: CARD_HEIGHT,
    width: "100%",
    borderRadius: 8,
  },
  infoContainer: {
    flex: 1,
    padding: 8,
  },
  title: {
    flex: 1,
    fontSize: 16,
    color: "white",
    fontWeight: "bold",
  },
  status: {
    fontSize: 12,
    color: "white",
  },
  image: {
    height: CARD_HEIGHT,
    width: CARD_WIDTH,
    borderTopLeftRadius: 8,
    borderBottomLeftRadius: 8,
  },
});
