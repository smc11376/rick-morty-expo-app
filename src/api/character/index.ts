import { useGetApiService } from "@/libs/network";
import { CharacterNetType, GetAllCharactersNetType } from "@/api/character/NetTypes";
import { transformGetAllCharactersResponse, transformGetCharacterByIdResponse } from "@/api/character/ResposeUtils";
import { Character, ShortCharacter } from "@/domain-models/character";
import { useEffect, useMemo, useState } from "react";

type UseAllCharactersType = {
  characters: ShortCharacter[];
  nextPage: string | null | undefined;
  isLoading: boolean;
};

type UseCharacterType = {
  character: Character | undefined;
  isLoading: boolean;
};

export function useGetAllCharacters(newUrl: string | null | undefined): UseAllCharactersType {
  const [characters, setCharacters] = useState<ShortCharacter[]>([]);
  const { data, isLoading, isFetching, isRefetching } = useGetApiService<GetAllCharactersNetType>(
    newUrl || "/character",
  );
  const newCharacters = useMemo(() => transformGetAllCharactersResponse(data), [data?.results]);

  useEffect(() => {
    if (!isLoading && !isFetching && !isRefetching) {
      setCharacters(currentCharacters => {
        return [...new Set([...currentCharacters, ...newCharacters])];
      });
    }
  }, [newCharacters]);

  return {
    characters,
    isLoading: isLoading || isFetching || isRefetching,
    nextPage: data?.info.next,
  };
}

export function useGetCharacterById(id?: string): UseCharacterType {
  if (!id) throw new Error("Oops! There is no character to find!");

  const { data, isLoading, isFetching, isRefetching } = useGetApiService<CharacterNetType>(`/character/${id}`);
  const character = transformGetCharacterByIdResponse(data);
  return {
    character,
    isLoading: isLoading || isFetching || isRefetching,
  };
}
