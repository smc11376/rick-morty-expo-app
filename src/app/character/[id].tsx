import { ScrollView, StyleSheet, View } from "react-native";
import { useGlobalSearchParams, useNavigation } from "expo-router";
import { useGetCharacterById } from "@/api/character";
import { useEffect } from "react";
import Loader from "@/components/loader";
import Row from "@/app/character/components/row";
import InfoBlock from "@/app/character/components/info-block";
import CharacterImage from "@/app/character/components/character-image";

export default function CharacterDetailScreen() {
  const params = useGlobalSearchParams();
  const { character, isLoading } = useGetCharacterById(params.id as string);
  const navigation = useNavigation();

  useEffect(() => {
    navigation.setOptions({ title: params.characterName });
  }, [params.characterName]);

  return isLoading ? (
    <View style={styles.container}>
      <Loader />
    </View>
  ) : (
    <ScrollView style={styles.container} contentContainerStyle={styles.contentContainerStyle}>
      <CharacterImage uri={character?.image} />
      <Row>
        <InfoBlock label="Gender: " data={character?.gender} />
        <InfoBlock label="Status: " data={character?.status} />
      </Row>
      <Row>
        <InfoBlock label="Species: " data={character?.species} />
        {character?.type !== "" && <InfoBlock label="Type: " data={character?.type} />}
      </Row>
      <InfoBlock label="Number of episodes: " data={character?.numberOfEpisodes.toString()} />
      <InfoBlock label="Origin: " data={character?.origin.name} urlToOpen={character?.origin.url} />
      <InfoBlock label="Current location: " data={character?.location.name} urlToOpen={character?.location.url} />
    </ScrollView>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: "rgb(60, 62, 68)",
  },
  contentContainerStyle: {
    flexGrow: 1,
  },
  image: {
    width: "100%",
    aspectRatio: 1,
  },
});
