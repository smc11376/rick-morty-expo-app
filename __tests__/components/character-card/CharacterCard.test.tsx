import { fireEvent, render, screen } from "@testing-library/react-native";
import CharacterCard from "@/components/character-card";
import { ShortCharacter } from "@/domain-models/character";

describe("CharacterCard renders properly with:", () => {
  it("Alive status", () => {
    const onPressCharacterMock = jest.fn();
    const mockShortCharacter: ShortCharacter = {
      id: 1,
      name: "testing name",
      status: "Alive",
      image: "https://rickandmortyapi.com/api/character/avatar/400.jpeg",
    };
    render(<CharacterCard character={mockShortCharacter} onPressCharacter={onPressCharacterMock} />);

    expect(screen.toJSON()).toMatchSnapshot();
  });

  it("Dead status", () => {
    const onPressCharacterMock = jest.fn();
    const mockShortCharacter: ShortCharacter = {
      id: 1,
      name: "testing name",
      status: "Dead",
      image: "https://rickandmortyapi.com/api/character/avatar/400.jpeg",
    };
    render(<CharacterCard character={mockShortCharacter} onPressCharacter={onPressCharacterMock} />);

    expect(screen.toJSON()).toMatchSnapshot();
  });

  it("unknown status", () => {
    const onPressCharacterMock = jest.fn();
    const mockShortCharacter: ShortCharacter = {
      id: 1,
      name: "testing name",
      status: "unknown",
      image: "https://rickandmortyapi.com/api/character/avatar/400.jpeg",
    };
    render(<CharacterCard character={mockShortCharacter} onPressCharacter={onPressCharacterMock} />);

    expect(screen.toJSON()).toMatchSnapshot();
  });
});

describe("CharacterCard actions:", () => {
  it("press on wrapper", () => {
    const onPressCharacterMock = jest.fn();
    const mockShortCharacter: ShortCharacter = {
      id: 1,
      name: "testing name",
      status: "Alive",
      image: "https://rickandmortyapi.com/api/character/avatar/400.jpeg",
    };
    render(<CharacterCard character={mockShortCharacter} onPressCharacter={onPressCharacterMock} />);

    fireEvent.press(screen.getByTestId("character-card-wrapper"));
    expect(onPressCharacterMock).toHaveBeenCalled();
  });
});
