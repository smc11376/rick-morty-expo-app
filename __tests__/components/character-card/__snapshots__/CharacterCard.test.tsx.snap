// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharacterCard renders properly with: Alive status 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "opacity": 1,
      "paddingHorizontal": 24,
      "paddingVertical": 8,
    }
  }
  testID="character-card-wrapper"
>
  <View
    style={
      {
        "alignItems": "center",
        "backgroundColor": "rgb(60, 62, 68)",
        "borderRadius": 8,
        "flexDirection": "row",
        "minHeight": 80,
        "width": "100%",
      }
    }
  >
    <Image
      resizeMode="contain"
      source={
        {
          "uri": "https://rickandmortyapi.com/api/character/avatar/400.jpeg",
        }
      }
      style={
        {
          "borderBottomLeftRadius": 8,
          "borderTopLeftRadius": 8,
          "height": 80,
          "width": 80,
        }
      }
    />
    <View
      style={
        {
          "flex": 1,
          "padding": 8,
        }
      }
    >
      <Text
        style={
          {
            "color": "white",
            "flex": 1,
            "fontSize": 16,
            "fontWeight": "bold",
          }
        }
      >
        testing name
      </Text>
      <Text
        style={
          {
            "color": "white",
            "fontSize": 12,
          }
        }
      >
        Status: 
        <Text
          style={
            {
              "color": "green",
            }
          }
        >
          Alive
        </Text>
      </Text>
    </View>
  </View>
</View>
`;

exports[`CharacterCard renders properly with: Dead status 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "opacity": 1,
      "paddingHorizontal": 24,
      "paddingVertical": 8,
    }
  }
  testID="character-card-wrapper"
>
  <View
    style={
      {
        "alignItems": "center",
        "backgroundColor": "rgb(60, 62, 68)",
        "borderRadius": 8,
        "flexDirection": "row",
        "minHeight": 80,
        "width": "100%",
      }
    }
  >
    <Image
      resizeMode="contain"
      source={
        {
          "uri": "https://rickandmortyapi.com/api/character/avatar/400.jpeg",
        }
      }
      style={
        {
          "borderBottomLeftRadius": 8,
          "borderTopLeftRadius": 8,
          "height": 80,
          "width": 80,
        }
      }
    />
    <View
      style={
        {
          "flex": 1,
          "padding": 8,
        }
      }
    >
      <Text
        style={
          {
            "color": "white",
            "flex": 1,
            "fontSize": 16,
            "fontWeight": "bold",
          }
        }
      >
        testing name
      </Text>
      <Text
        style={
          {
            "color": "white",
            "fontSize": 12,
          }
        }
      >
        Status: 
        <Text
          style={
            {
              "color": "red",
            }
          }
        >
          Dead
        </Text>
      </Text>
    </View>
  </View>
</View>
`;

exports[`CharacterCard renders properly with: unknown status 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "opacity": 1,
      "paddingHorizontal": 24,
      "paddingVertical": 8,
    }
  }
  testID="character-card-wrapper"
>
  <View
    style={
      {
        "alignItems": "center",
        "backgroundColor": "rgb(60, 62, 68)",
        "borderRadius": 8,
        "flexDirection": "row",
        "minHeight": 80,
        "width": "100%",
      }
    }
  >
    <Image
      resizeMode="contain"
      source={
        {
          "uri": "https://rickandmortyapi.com/api/character/avatar/400.jpeg",
        }
      }
      style={
        {
          "borderBottomLeftRadius": 8,
          "borderTopLeftRadius": 8,
          "height": 80,
          "width": 80,
        }
      }
    />
    <View
      style={
        {
          "flex": 1,
          "padding": 8,
        }
      }
    >
      <Text
        style={
          {
            "color": "white",
            "flex": 1,
            "fontSize": 16,
            "fontWeight": "bold",
          }
        }
      >
        testing name
      </Text>
      <Text
        style={
          {
            "color": "white",
            "fontSize": 12,
          }
        }
      >
        Status: 
        <Text
          style={
            {
              "color": "grey",
            }
          }
        >
          unknown
        </Text>
      </Text>
    </View>
  </View>
</View>
`;
